File: JAVA/101 - Symmetric Tree.java
/*
 101. Symmetric Tree (E)

 Approach 1 : Preorder BFS
 Time: O(n)
 Space: O(n)

 Approach 2 : Recursion DFS
 Time: O(n)
 Space:O(n), n is the depth of the recursion

*/

/* Approach 1 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        Queue<TreeNode> todo = new LinkedList<TreeNode>();
        todo.add(root.left);
        todo.add(root.right);
        while (!todo.isEmpty()) {
            TreeNode p = todo.poll();
            TreeNode q = todo.poll();
            
            if (p == null && q == null) continue;
            if (p == null || q == null) return false;
            if (p.val != q.val) return false;
            
            todo.add(p.left);
            todo.add(q.right);
            todo.add(p.right);
            todo.add(q.left);
        }
        return true;
    }
}


/* Approach 2 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return isSymmetric(root.left, root.right);
    }
    
    public boolean isSymmetric(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;
        if (p.val != q.val) return false;
        return isSymmetric(p.left, q.right) && isSymmetric(p.right, q.left);
    }
}

File: C++/101-Symmetric_Tree.cpp
/*
 101. Symmetric Tree (E)

 BFS
 O(n),O(n)

 DFS
 O(n),O(n)

*/

// BFS
class Solution
{
public:
    bool isSymmetric(TreeNode *root)
    {
        if (!root) return true;

        // Use queue to record the nodes that will be processed
        queue<TreeNode *> todo;
        todo.push(root->left);
        todo.push(root->right);

        TreeNode *p = nullptr;
        TreeNode *q = nullptr;

        while (!todo.empty())
        {
            p = todo.front();
            todo.pop();
            q = todo.front();
            todo.pop();

            if (!p && !q) continue;
            if (!p || !q) return false;
            if (p->val != q->val) return false;

            todo.push(p->left);
            todo.push(q->right);
            todo.push(p->right);
            todo.push(q->left);
        }
        return true;
    }
};

// DFS
class Solution
{
public:
    bool isSymmetric(TreeNode *root)
    {
        if (!root) return true;
        return Check(root->left, root->right);
    }

    bool Check(TreeNode *l, TreeNode *r)
    {
        if (!l && !r) return true;
        if (!l || !r) return false;
        if (l->val != r->val) return false;
        return Check(l->left, r->right) && Check(l->right, r->left);
    }
};
